core/src/main/java/net/adoptopenjdk/icedteaweb/resources/downloader/PrioritizedParallelExecutor.java

```diff
@@ -0,0 +1,220 @@
+package net.adoptopenjdk.icedteaweb.resources.downloader;
+
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+/**
+ * Executor which will execute multiple {@link Callable Callables} in parallel.
+ * The callables must be sorted and have the highest priority callable as the first in list.
+ * Executions of lower priority are canceled as soon as a higher priority result has been successfully been returned.
+ * The final {@link Future} will complete as soon as the first callable completed successfully and all other callables
+ * with higher priority (if any) have completed exceptionally
+ */
+public class PrioritizedParallelExecutor {
+
+    private final ExecutorService executor;
+
+    /**
+     * Constructor
+     *
+     * @param executor the executor service which is used for executing the callables.
+     */
+    public PrioritizedParallelExecutor(ExecutorService executor) {
+        this.executor = executor;
+    }
+
+    /**
+     * Executes the passed callables in parallel
+     * The callables must be sorted by priority with the highest priority as the first in list.
+     * Executions of lower priority are canceled as soon as a higher priority result has successfully been calculated.
+     * The returned {@link Future} will complete as soon as the first callable completed successfully and all other
+     * callables with higher priority (if any) have completed exceptionally.
+     *
+     * @param callables the callables to execute.
+     * @return a future holding the best result from all callables.
+     */
+    public <V> Future<V> getSuccessfulResultWithHighestPriority(List<Callable<V>> callables) {
+        if (callables == null || callables.isEmpty()) {
+            throw new IllegalArgumentException("No callables");
+        }
+
+        if (callables.size() == 1) {
+            return executor.submit(callables.get(0));
+        }
+
+        final CancelingCallable<V> head = new CancelingCallable<>(callables.get(0));
+        final List<Callable<V>> tail = callables.subList(1, callables.size());
+
+        final ChainableFuture<V> result = new ChainableFuture<>(executor.submit(head));
```

Comment:
This looks like you pass each Callable 2 times to the executor. First you add it in line 54 and now you add it a second time. So for 4 callable in the list we need 8 tasks runs in the executor.

Reply:
no, I only see this one submit, there is not other one

Reply:
line 54 and line 60

Reply:
but they are exclusive.
Line 54:  `return executor.submit(callables.get(0));`

So if the callable is submitted there it will never be submitted on line 60 and vice-versa

Reply:
ok, missed the `return`. My fault ;)